drivers/mbox/mbox_stm32_hsem.c

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (c) 2024 Celina Sophie Kalus <hello@celinakalus.de>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/mbox.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+
+#include "stm32_hsem.h"
+
+#define LOG_LEVEL CONFIG_MBOX_LOG_LEVEL
+#include <zephyr/logging/log.h>
+#include <zephyr/irq.h>
+LOG_MODULE_REGISTER(mbox_stm32_hsem_ipc);
+
+#define DT_DRV_COMPAT st_mbox_stm32_hsem
+
+#define HSEM_CPU1                   1
+#define HSEM_CPU2                   2
+
+#if DT_NODE_EXISTS(DT_NODELABEL(cpu0))
+#define HSEM_CPU_ID HSEM_CPU1
+#elif DT_NODE_EXISTS(DT_NODELABEL(cpu1))
+#define HSEM_CPU_ID HSEM_CPU2
+#else
+#error "Neither cpu0 nor cpu1 defined!"
+#endif
+
+#if HSEM_CPU_ID == HSEM_CPU1
+#define ll_hsem_enableit_cier       LL_HSEM_EnableIT_C1IER
+#define ll_hsem_disableit_cier      LL_HSEM_DisableIT_C1IER
+#define ll_hsem_clearflag_cicr      LL_HSEM_ClearFlag_C1ICR
+#define ll_hsem_isactiveflag_cmisr   LL_HSEM_IsActiveFlag_C1MISR
+#else /* HSEM_CPU2 */
+#define ll_hsem_enableit_cier       LL_HSEM_EnableIT_C2IER
+#define ll_hsem_disableit_cier      LL_HSEM_DisableIT_C2IER
+#define ll_hsem_clearflag_cicr      LL_HSEM_ClearFlag_C2ICR
+#define ll_hsem_isactiveflag_cmisr   LL_HSEM_IsActiveFlag_C2MISR
+#endif /* HSEM_CPU_ID */
+
+#define MBOX_HSEM_TX_MASK(idx, _) BIT(DT_INST_PROP_BY_IDX(0, st_hsem_ids_tx, idx))
+#define MBOX_HSEM_RX_MASK(idx, _) BIT(DT_INST_PROP_BY_IDX(0, st_hsem_ids_rx, idx))
+
+#define MAX_TX_CHANNELS DT_INST_PROP_LEN(0, st_hsem_ids_tx)
+#define MAX_RX_CHANNELS DT_INST_PROP_LEN(0, st_hsem_ids_rx)
+#define MAX_CHANNELS (MAX_RX_CHANNELS + MAX_TX_CHANNELS)
+
+
+struct mbox_stm32_hsem_data {
+	const struct device *dev;
+	mbox_callback_t cb[MAX_RX_CHANNELS];
+	void *user_data[MAX_RX_CHANNELS];
+};
+
+static struct mbox_stm32_hsem_data stm32_hsem_mbox_data;
+
+
+static struct mbox_stm32_hsem_conf {
+	struct stm32_pclken pclken;
+	int rx_hsem_ids[MAX_RX_CHANNELS];
+	int tx_ch_mask;
+	int rx_ch_mask;
+} stm32_hsem_mbox_conf = {
+	.tx_ch_mask = LISTIFY(MAX_TX_CHANNELS, MBOX_HSEM_TX_MASK, |),
+	.rx_ch_mask = LISTIFY(MAX_RX_CHANNELS, MBOX_HSEM_RX_MASK, |),
+	.rx_hsem_ids = DT_INST_PROP(0, st_hsem_ids_rx),
+	.pclken = {
+		.bus = DT_INST_CLOCKS_CELL(0, bus),
+		.enr = DT_INST_CLOCKS_CELL(0, bits)
+	},
+};
+
+static inline bool is_rx_channel_valid(const struct device *dev, uint32_t ch)
+{
+	const struct mbox_stm32_hsem_conf *cfg = dev->config;
+
+	return (cfg->rx_ch_mask & BIT(ch));
```

Comment:
After discussion with @arnopo :
 Go with "Implement a test using [CFG_HW_IPM_CPU1_SEMID](https://elixir.bootlin.com/zephyr/latest/C/ident/CFG_HW_IPM_CPU1_SEMID) or [CFG_HW_IPM_CPU2_SEMID](https://elixir.bootlin.com/zephyr/latest/C/ident/CFG_HW_IPM_CPU2_SEMID), depending on the HSEM_CPU_ID":
- remove properties:  st,hsem_ids_tx/rx
- Implement check vs header in current function.

Ideally, we could make it more simple by moving definitions from stm32_hsem.h to actual device tree headers that could be included in dt and driver and remove the need for a check. Not requested today since this would require additional unrelated work (split headers, ..).

Reply:
I am not sure how to understand this. Do I
1. allow ONLY the same IPM semaphores, OR
2. allow arbitrary semaphore IDs ABOVE those IPM semaphores?

Reply:
@erwango @arnopo I hope you are doing well. Could you give me some additional clarification about how I should be implementing this? See my previous comment. Thank you in advance!

Reply:
I'd say allow only the same IPM semaphores